### PR TITLE
Create unique constraint on managed book & text to avoid creating obvious duplicates

### PR DESCRIPTION
**Pull request by z1lc**

### CODE REVIEW

File: app/readwise.py [hunk 1]
```diff
@@ -47,7 +47,17 @@ def upsert_highlight(user: User, highlight: JsonDict):
     maybe_highlight.id = id
     maybe_highlight.managed_readwise_book_id = corresponding_managed_book_id
     maybe_highlight.text = highlight["text"]
-    db.session.commit()
+    try:
+        db.session.commit()
+    except Exception as e:
+        # If the exception is because of the unique constraint, it's fine, just rollback + skip.
+        # If it's not, though, we'd like to know.
+        if '_managed_readwise_book_id_and_text' in repr(e):
```
Comment (wclausen) [inline app/readwise.py:55]: Something to potentially add to the backlog, matching on string representation is not ideal for error handling. Ideally, we could rely on something that is part of the public api from sqlalchemy (like an exception type). Not blocking rn, just something that is not ideal

File: app/routes.py [hunk 2]
```diff
@@ -16,9 +16,9 @@
 from werkzeug.urls import url_parse
 
 from app.card_generation.anki import generate_full_apkg
-from app.models.base import User
 from app.card_generation.util import AnkiCard
 from app.models.anki import AnkiReviewLog
+from app.models.base import User
```
Comment (wclausen) [inline app/routes.py:21]: This seems weird. Maybe a holdover from testing? Do we need this?

File: tests/test_readwise.py [hunk 3]
```diff
@@ -1,9 +1,15 @@
...
-from app.readwise import refresh_highlights_and_books
+from app.readwise import refresh_highlights_and_books, get_paginated
 
 
 @pytest.mark.skip(reason="integration")
 def test_refresh_highlights_and_books():
     refresh_highlights_and_books(User.query.filter_by(username='rsanek').one())
+
+
+@pytest.mark.skip(reason="integration")
```
Comment (wclausen) [inline tests/test_readwise.py:12]: i like this. would it be a good thing to add to the backlog to set up daily runs of these integration tests?

**Review by wclausen — APPROVED** [review]

File: app/readwise.py (hunk 1 above)
Comment (z1lc) [reply app/readwise.py:55]: Yeah there is an `IntegrityError` that I can use but it's not specific to the uniqueness constraint -- using this string means we know *for sure* that the constraint is what we're catching.

File: app/routes.py (hunk 2 above)
Comment (z1lc) [reply app/routes.py:21]: It's just an ordering thing -- `app.models.anki ` comes before `app.models.base`. It's just moving this line lower

File: tests/test_readwise.py (hunk 3 above)
Comment (z1lc) [reply tests/test_readwise.py:12]: https://github.com/z1lc/zdone/issues/115

**Pull request merged**